### PR TITLE
Tesla Counterplay.

### PR DESCRIPTION
**Pull request by jack-avery**
## About The Pull Request

- Grounding rods now pull in the Tesla. The nearest grounding rod is prioritized.
- If the Tesla has one or more miniballs, the grounding rod will be consumed and take some of the Tesla's energy with it (equivalent to two miniballs)
- If the Tesla has no miniballs (energy of 12 or below), it will continue to be anchored nearby a grounding rod (if there is one).
- Teslas will now fizzle once their energy reaches 0. There is an admin log for this.
- The Tesla 'contained' variable now functions as expected, with it's own unique check. Works guaranteed with containments up to 9x9 tiles of space (guaranteed on roundstart setups for Delta and Lambda stations).

## Why It's Good For The Game

tesla counterplay

## Changelog
:cl:
tweak: tesla movement priorities
add: tesla counterplay
add: tesla containment check (containment variable now usable)
/:cl:

<details>
  <summary>old description</summary>
  Grounding rods now attract the Tesla. Exactly how much is luck-based, but you can expect it to gravitate generally towards a rod.
Grounding rods, when zapped by a Tesla with miniballs, will eat a chunk of energy from the Tesla equivalent to two miniballs.
This gives a Tesloose some considerable counterplay, and since this is unmapped territory, we'll have to see whether this needs buffing or not.
Testmerge suggested to see if this is powerful enough before a fullmerge, though the situations under which this occurs are generally rare.

Edit, for more details on the effects:
A Tesla, when surrounded by containment fields on all sides within a radius 10 tiles, is considered contained. If your Tesla containment is 10x10 (or smaller) this is fine (and this is how I most often see the Tesla contained). But, if the Tesla containment is larger than 10x10, the Tesla has a decent chance to move out of range and be considered uncontained during this tick, destroying one (or more) of the nearby grounding rods, opening up the rest of the area to be fried.
</details>

### CODE REVIEW

**Comment by Detective-Google** [comment]
fun fact teslas don't die without any powersource; the main ball will never fade

**Comment by MrJWhit** [comment]
It will sit very politely on the final grounding rod then.

File: code/modules/power/tesla/energy_ball.dm [hunk 1]
```diff
@@ -31,6 +31,7 @@
 	var/produced_power
 	var/energy_to_raise = 32
 	var/energy_to_lower = -20
+	var/rodtarget
```
Comment (Putnam3145) [inline code/modules/power/tesla/energy_ball.dm]:
```suggestion
	var/obj/machinery/power/grounding_rod/rodtarget
```

**Comment by jack-avery** [comment]
i know the ball doesn't die, but it will stay somewhat near the grounding rod and a majority of the shocking from a low tier ball will continue to be eaten up by the rod

File: code/modules/power/tesla/energy_ball.dm [hunk 2]
```diff
@@ -91,8 +92,21 @@
 /obj/singularity/energy_ball/proc/move_the_basket_ball(var/move_amount)
 	//we face the last thing we zapped, so this lets us favor that direction a bit
 	var/move_bias = pick(GLOB.alldirs)
+	var/move_dir
+	var/rods = GLOB.grounding_rods
+	if(rods) // grounding rods pull the tesla ball, picks the nearest one
```
Comment (Putnam3145) [inline code/modules/power/tesla/energy_ball.dm]: this is unnecessary, if GLOB doesn't have grounding_rods it'll yell anyway afaik

File: code/modules/power/tesla/energy_ball.dm [hunk 3]
```diff
@@ -91,8 +92,21 @@
...
+		for(var/rod in rods) 
+			if(!rodtarget)
```
Comment (Putnam3145) [inline code/modules/power/tesla/energy_ball.dm]:
```suggestion
			if(!rodtarget || get_dist(src,rod)<get_dist(src,rodtarget))
```

File: code/modules/power/tesla/energy_ball.dm [hunk 4]
```diff
@@ -91,8 +92,21 @@
...
+				rodtarget=rod
+			
+			if(get_dist(src,rod)<get_dist(src,rodtarget))
+				rodtarget=rod
```
Comment (Putnam3145) [inline code/modules/power/tesla/energy_ball.dm]: if committing above suggestion, remove these two lines

File: code/modules/power/tesla/energy_ball.dm [hunk 5]
```diff
@@ -91,8 +92,21 @@
...
+
 	for(var/i in 0 to move_amount)
-		var/move_dir = pick(GLOB.alldirs + move_bias) //ensures large-ball teslas don't just sit around
+		if(rods)
```
Comment (Putnam3145) [inline code/modules/power/tesla/energy_ball.dm]:
```suggestion
		if(rodtarget)
```

**Review by Putnam3145 — CHANGES_REQUESTED** [review]

**Comment by timothyteakettle** [comment]
this is a good start

File: code/modules/power/tesla/energy_ball.dm (hunk 2 above)
Comment (jack-avery) [reply code/modules/power/tesla/energy_ball.dm]: i did this for readability more than anything, would removing the if(rods) and replacing the for(in rods) with (in GLOB) work?

**Comment by jack-avery** [comment]
apparently still deletes grounding rods while contained; working on a fix hold up

**Comment by jack-avery** [comment]
wrong button

**Comment by jack-avery** [comment]
^ only works with 1x1s, working on a fix for larger containment areas

**Comment by Putnam3145** [comment]
roundstart setups are always more than 1x1 and lambda still has a roundstart setup (though, to be fair, I don't want it to)

**Comment by raspy-on-osu** [comment]
until somehow i figure out how to better determine when it's contained im gonna shelf this, unless people are understanding of this side-effect which can _actually be quite dangerous_

**Comment by YakumoChen** [comment]
This kills my infamous Multiball setup, please no kill

**Comment by raspy-on-osu** [comment]
just contain each of them separately in a grid?

**Comment by lolman360** [comment]
@raspy-on-osu maybe have it check for containment fields on all 4 sides occasionally to decide "containment"

not just one in range because you could probably use it to cheese a tesloose

**Comment by jack-avery** [comment]
planned to do this, will try now that i'm reminded of it

**Comment by jack-avery** [comment]
ready; cleaned up cluttery comments

**Review by Putnam3145 — APPROVED** [review]

**Pull request merged**